File: backend/src/repositories/product.repository.js
const pool = require('../config/database');

class ProductRepository {
  async findAll({ vendorId, categoryId } = {}) {
    let query = `
      SELECT p.*, u.name as vendor_name, c.name as category_name
      FROM products p
      JOIN users u ON p.vendor_id = u.id
      LEFT JOIN categories c ON p.category_id = c.id
      WHERE p.stock > 0
    `;
    const params = [];
    if (vendorId) {
      params.push(vendorId);
      query += ` AND p.vendor_id = $${params.length}`;
    }
    if (categoryId) {
      params.push(categoryId);
      query += ` AND p.category_id = $${params.length}`;
    }
    query += ' ORDER BY p.created_at DESC';
    const result = await pool.query(query, params);
    return result.rows;
  }

  async findById(id) {
    const result = await pool.query(
      `SELECT p.*, u.name as vendor_name, c.name as category_name
       FROM products p
       JOIN users u ON p.vendor_id = u.id
       LEFT JOIN categories c ON p.category_id = c.id
       WHERE p.id = $1`,
      [id]
    );
    return result.rows[0] || null;
  }

  async create({ name, description, price, stock, vendorId, categoryId }) {
    const result = await pool.query(
      `INSERT INTO products (name, description, price, stock, vendor_id, category_id)
       VALUES ($1, $2, $3, $4, $5, $6)
       RETURNING *`,
      [name, description, price, stock, vendorId, categoryId]
    );
    return result.rows[0];
  }

  async update(id, { name, description, price, stock, categoryId }) {
    const result = await pool.query(
      `UPDATE products
       SET name = $1, description = $2, price = $3, stock = $4, category_id = $5, updated_at = NOW()
       WHERE id = $6 RETURNING *`,
      [name, description, price, stock, categoryId, id]
    );
    return result.rows[0] || null;
  }

  // Décrémentation de stock avec verrou (gestion concurrence)
  async decrementStock(id, quantity, client) {
    const db = client || pool;
    const result = await db.query(
      `UPDATE products
       SET stock = stock - $1
       WHERE id = $2 AND stock >= $1
       RETURNING id, stock`,
      [quantity, id]
    );
    return result.rows[0] || null;
  }

  async delete(id) {
    await pool.query('DELETE FROM products WHERE id = $1', [id]);
  }

  async findCategories() {
    const result = await pool.query('SELECT * FROM categories ORDER BY name');
    return result.rows;
  }
}

module.exports = new ProductRepository();

File: backend/repositories/product_repository.py
from typing import Optional
from sqlalchemy.orm import Session
from sqlalchemy import func
from models.product import Product, Category
from models.user import User


class ProductRepository:

    def find_all(self, db: Session, vendor_id: Optional[str] = None, category_id: Optional[str] = None):
        query = (
            db.query(Product, User.name.label("vendor_name"), Category.name.label("category_name"))
            .join(User, Product.vendor_id == User.id)
            .outerjoin(Category, Product.category_id == Category.id)
            .filter(Product.stock > 0)
        )
        if vendor_id:
            query = query.filter(Product.vendor_id == vendor_id)
        if category_id:
            query = query.filter(Product.category_id == category_id)
        rows = query.order_by(Product.created_at.desc()).all()
        return [self._merge(p, vendor_name, category_name) for p, vendor_name, category_name in rows]

    def find_by_id(self, db: Session, product_id: str):
        row = (
            db.query(Product, User.name.label("vendor_name"), Category.name.label("category_name"))
            .join(User, Product.vendor_id == User.id)
            .outerjoin(Category, Product.category_id == Category.id)
            .filter(Product.id == product_id)
            .first()
        )
        if not row:
            return None
        return self._merge(row[0], row[1], row[2])

    def create(self, db: Session, name: str, description: Optional[str], price, stock: int,
               vendor_id: str, category_id: Optional[str]) -> Product:
        product = Product(
            name=name,
            description=description,
            price=price,
            stock=stock,
            vendor_id=vendor_id,
            category_id=category_id,
        )
        db.add(product)
        db.commit()
        db.refresh(product)
        return product

    def update(self, db: Session, product_id: str, **kwargs) -> Optional[Product]:
        product = db.query(Product).filter(Product.id == product_id).first()
        if not product:
            return None
        for key, value in kwargs.items():
            if value is not None:
                setattr(product, key, value)
        db.commit()
        db.refresh(product)
        return product

    def decrement_stock(self, db: Session, product_id: str, quantity: int) -> Optional[Product]:
        """Décrémentation atomique du stock (concurrent safe avec with_for_update)."""
        product = (
            db.query(Product)
            .filter(Product.id == product_id, Product.stock >= quantity)
            .with_for_update()
            .first()
        )
        if not product:
            return None
        product.stock -= quantity
        db.flush()
        return product

    def delete(self, db: Session, product_id: str):
        db.query(Product).filter(Product.id == product_id).delete()
        db.commit()

    def find_categories(self, db: Session):
        return db.query(Category).order_by(Category.name).all()

    @staticmethod
    def _merge(product: Product, vendor_name: str, category_name: Optional[str]) -> dict:
        """Retourne un dict avec les champs du produit + vendor_name + category_name."""
        data = {c.name: getattr(product, c.name) for c in product.__table__.columns}
        data["vendor_name"] = vendor_name
        data["category_name"] = category_name
        return data


product_repo = ProductRepository()
